### PR TITLE
[PET-to-MLIR] createFor - added symbolic bounds

### PR DESCRIPTION
**Pull request by Komisarczyk**
Signed-off-by: Komisarczyk <konrad.komisarczyk@gmail.com>

### CODE REVIEW

**Comment by Komisarczyk** [comment]
is the style of if's in createfor in IslNodeBuilder.cc ok?

I forgot that casting the char* to string is unnecessary, if you prefer that changed and tested now let me know

File: mlir/tools/mlir-pet/islNodeBuilder.cc [hunk 1]
```diff
@@ -29,19 +29,72 @@ static int createIntFromIslExpr(isl::ast_expr expression) {
 // TODO: See how we can get location information.
 // TODO: handle degenerate loop (see isl_ast_node_for_is_degenerate)
 // TODO: See how to handle more complex expression in the loop.
+
+bool isInt(isl::ast_expr expression) {
```
Comment (chelini) [inline mlir/tools/mlir-pet/islNodeBuilder.cc:33]: Please remove "if". Just write return isl_ast_expr_get_type(expression.get()) == isl_ast_expr_in in the body of the function.

File: mlir/tools/mlir-pet/islNodeBuilder.cc [hunk 2]
```diff
@@ -29,19 +29,72 @@ static int createIntFromIslExpr(isl::ast_expr expression) {
...
+  if (isl_ast_expr_get_type(expression.get()) == isl_ast_expr_int)
+    return true;
+  return false;
+}
+bool isId(isl::ast_expr expression) {
```
Comment (chelini) [inline mlir/tools/mlir-pet/islNodeBuilder.cc:36]: Same here.

File: mlir/tools/mlir-pet/islNodeBuilder.cc [hunk 3]
```diff
@@ -29,19 +29,72 @@ static int createIntFromIslExpr(isl::ast_expr expression) {
...
+  if (isl_ast_expr_get_type(expression.get()) == isl_ast_expr_id)
+    return true;
+  return false;
+}
 void IslNodeBuilder::createFor(isl::ast_node forNode) {
-  auto lowerBound = forNode.for_get_init();
```
Comment (chelini) [inline mlir/tools/mlir-pet/islNodeBuilder.cc:33]: Is this related to your changes? Or you just move it?

File: mlir/tools/mlir-pet/islNodeBuilder.cc [hunk 4]
```diff
@@ -29,19 +29,72 @@ static int createIntFromIslExpr(isl::ast_expr expression) {
...
   auto increment = forNode.for_get_inc();
   auto iterator = forNode.for_get_iterator();
   auto iteratorId = iterator.get_id().to_str();
   auto upperBound = getUpperBound(forNode);
+  auto lowerBound = forNode.for_get_init();
+  auto incrementAsInt = createIntFromIslExpr(increment);
```
Comment (chelini) [inline mlir/tools/mlir-pet/islNodeBuilder.cc]: Did you rebase on the last version? Now we have something like: auto lowerbound = std::abs(/*...*/);

File: mlir/tools/mlir-pet/islNodeBuilder.cc [hunk 5]
```diff
@@ -29,19 +29,72 @@ static int createIntFromIslExpr(isl::ast_expr expression) {
...
+  AffineForOp loop;
+  // symbolic upper bound
+  if (isInt(lowerBound) && isId(upperBound)) {
+    // get id from node type
+    auto upperBound_id = isl_ast_expr_get_id(upperBound.get());
```
Comment (chelini) [inline mlir/tools/mlir-pet/islNodeBuilder.cc]: Use camelCase (i.e., upperBound_id -> upperBoundId) [this is the style used in LLVM, better to be consistent with it.]

File: mlir/tools/mlir-pet/islNodeBuilder.cc [hunk 6]
```diff
@@ -29,19 +29,72 @@ static int createIntFromIslExpr(isl::ast_expr expression) {
...
+    // access string name
+    auto upperBound_processed = isl_id_get_name(upperBound_id);
+
+    auto lowerBound_processed = createIntFromIslExpr(lowerBound);
+    loop = MLIRBuilder_.createLoop(lowerBound_processed, upperBound_processed,
+                                   incrementAsInt);
+    isl_id_free(upperBound_id);
+  }
 
-  auto lowerBoundAsInt = std::abs(createIntFromIslExpr(lowerBound));
```
Comment (chelini) [inline mlir/tools/mlir-pet/islNodeBuilder.cc:39]: The std::abs(was introduced for handling loops going downward. Did you have any test failures?

File: mlir/tools/mlir-pet/mlirCodegen.cc [hunk 7]
```diff
@@ -515,13 +515,83 @@ LogicalResult MLIRCodegen::verifyModule() {
...
+  builder_.setInsertionPointToStart(loop.getBody());
+
+  return loop;
+}
+
+AffineForOp MLIRCodegen::createLoop(std::string lb_id, std::string ub_id,
+                                    int step) {
+  Value ub, lb;
+  if (failed(this->getLoopTable().find(ub_id, ub)))
+    outs() << "failed";
```
Comment (chelini) [inline mlir/tools/mlir-pet/mlirCodegen.cc]: llvm_unrechable?

**Comment by chelini** [comment]
What do you mean by style? As in LLVM we use camelCase.

File: mlir/tools/mlir-pet/islNodeBuilder.cc (hunk 3 above)
Comment (Komisarczyk) [reply mlir/tools/mlir-pet/islNodeBuilder.cc:33]: just grouped it with upperbound
